clippy_lints/src/slice_as_bytes.rs

```diff
@@ -0,0 +1,58 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, source::snippet_with_applicability, ty::is_type_lang_item};
+use rustc_errors::Applicability::MachineApplicable;
+use rustc_hir::{Expr, ExprKind, LangItem, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for string slices immediantly followed by `as_bytes`.
+    /// ### Why is this bad?
+    /// This involves doing an unnecessary UTF-8 alignment check which is less efficient, and can cause a panic.
+    /// ### Example
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// s[1..5].as_bytes();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let s = "Lorem ipsum";
+    /// &s.as_bytes()[1..5];
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub SLICE_AS_BYTES,
+    nursery,
+    "slicing a string and immediantly calling as_bytes is less efficient and can lead to panics"
+}
+declare_lint_pass!(SliceAsBytes => [SLICE_AS_BYTES]);
+
+impl<'tcx> LateLintPass<'tcx> for SliceAsBytes {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if let ExprKind::MethodCall(path, recv, [], _) = expr.kind && path.ident.name == sym!(as_bytes){
```

Comment:
I would probably change this `if` statement order (it's slightly unoptimal). Also, knowing that `as_bytes` always always takes 0 arguments, you can just ignore the third and fourth arguments (using `..`)

```suggestion
        if path.ident.name == sym!(as_bytes) && let ExprKind::MethodCall(path, recv, ..) = expr.kind {
```